File: skills/code-architect/references/code-examples/python/prototype.py
"""Prototype — copy existing objects via a clone interface without coupling to classes."""

from __future__ import annotations

import copy
from abc import ABC, abstractmethod


class Shape(ABC):
    def __init__(self, color: str, tags: list[str] | None = None) -> None:
        self.color = color
        self.tags = tags or []

    @abstractmethod
    def clone(self) -> "Shape":
        ...

    def __str__(self) -> str:
        return f"{type(self).__name__}(color={self.color!r}, tags={self.tags})"


class Circle(Shape):
    def __init__(self, color: str, radius: float, tags: list[str] | None = None) -> None:
        super().__init__(color, tags)
        self.radius = radius

    def clone(self) -> "Circle":
        return copy.deepcopy(self)


class Rectangle(Shape):
    def __init__(self, color: str, w: float, h: float, tags: list[str] | None = None) -> None:
        super().__init__(color, tags)
        self.w, self.h = w, h

    def clone(self) -> "Rectangle":
        return copy.deepcopy(self)


if __name__ == "__main__":
    original = Circle("red", 4.0, tags=["base"])
    copy_ = original.clone()
    copy_.color = "blue"
    copy_.tags.append("derived")
    print("original:", original)
    print("clone:   ", copy_)

File: skills/code-architect/references/code-examples/ruby/prototype.rb
# frozen_string_literal: true

# Prototype — copy existing objects via a clone method.

# Prototype role: knows how to produce a deep, independent copy of itself.
class Shape
  attr_accessor :origin, :tags

  def initialize(origin:, tags: [])
    @origin = origin   # mutable [x, y]
    @tags = tags       # mutable list
  end

  # Custom deep copy so clones don't share mutable internals.
  def initialize_copy(_source)
    super
    @origin = @origin.dup
    @tags = @tags.dup
  end

  def to_s = "#{self.class.name}(origin=#{@origin.inspect}, tags=#{@tags.inspect})"
end

class Circle < Shape
  attr_accessor :radius

  def initialize(origin:, radius:, tags: [])
    super(origin: origin, tags: tags)
    @radius = radius
  end

  def to_s = "Circle(origin=#{@origin.inspect}, r=#{@radius}, tags=#{@tags.inspect})"
end

if __FILE__ == $PROGRAM_NAME
  original = Circle.new(origin: [0, 0], radius: 5, tags: ["base"])
  copy = original.clone

  copy.origin[0] = 99
  copy.tags << "variant"
  copy.radius = 10

  puts "original: #{original}"
  puts "copy:     #{copy}"
end
